src/Functions/mortonDecode.cpp

```diff
@@ -0,0 +1,427 @@
+            tuple_columns[2] = std::move(col2); \
+            tuple_columns[3] = std::move(col3); \
+            tuple_columns[4] = std::move(col4); \
+            tuple_columns[5] = std::move(col5); \
+            return ColumnTuple::create(tuple_columns); \
+        case 7: \
+            tuple_columns[0] = std::move(col0); \
+            tuple_columns[1] = std::move(col1); \
+            tuple_columns[2] = std::move(col2); \
+            tuple_columns[3] = std::move(col3); \
+            tuple_columns[4] = std::move(col4); \
+            tuple_columns[5] = std::move(col5); \
+            tuple_columns[6] = std::move(col6); \
+            return ColumnTuple::create(tuple_columns); \
+        case 8: \
+            tuple_columns[0] = std::move(col0); \
+            tuple_columns[1] = std::move(col1); \
+            tuple_columns[2] = std::move(col2); \
+            tuple_columns[3] = std::move(col3); \
+            tuple_columns[4] = std::move(col4); \
+            tuple_columns[5] = std::move(col5); \
+            tuple_columns[6] = std::move(col6); \
+            tuple_columns[7] = std::move(col7); \
+            return ColumnTuple::create(tuple_columns); \
+    } \
+    return ColumnTuple::create(tuple_columns);
+
+DECLARE_DEFAULT_CODE(
+constexpr auto MortonND_2D_Dec = mortonnd::MortonNDLutDecoder<2, 32, 8>();
+constexpr auto MortonND_3D_Dec = mortonnd::MortonNDLutDecoder<3, 21, 8>();
+constexpr auto MortonND_4D_Dec = mortonnd::MortonNDLutDecoder<4, 16, 8>();
+constexpr auto MortonND_5D_Dec = mortonnd::MortonNDLutDecoder<5, 12, 8>();
+constexpr auto MortonND_6D_Dec = mortonnd::MortonNDLutDecoder<6, 10, 8>();
+constexpr auto MortonND_7D_Dec = mortonnd::MortonNDLutDecoder<7, 9, 8>();
+constexpr auto MortonND_8D_Dec = mortonnd::MortonNDLutDecoder<8, 8, 8>();
+class FunctionMortonDecode : public IFunction
+{
+public:
+    static constexpr auto name = "mortonDecode";
+    static FunctionPtr create(ContextPtr)
+    {
+        return std::make_shared<FunctionMortonDecode>();
+    }
+
+    String getName() const override
+    {
+        return name;
+    }
+
+    size_t getNumberOfArguments() const override
+    {
+        return 2;
+    }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+    ColumnNumbers getArgumentsThatAreAlwaysConstant() const override { return {0}; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        UInt64 tuple_size = 0;
+        const auto * col_const = typeid_cast<const ColumnConst *>(arguments[0].column.get());
+        if (col_const)
+        {
+            const auto * mask = typeid_cast<const ColumnTuple *>(col_const->getDataColumnPtr().get());
+            if (mask)
+            {
+                tuple_size = mask->tupleSize();
+                for (size_t i = 0; i < tuple_size; i++)
+                {
+                    auto ratio = mask->getColumn(i).getUInt(0);
```

Comment:
Here we probably need to check that our column is actually a number and not a string or a tuple :)

Reply:
Yup, but that's not what was problematic.
I have refactored all checks to not use `column->getName()` and it immediately fixed the problem. So it looks like `getName` cannot be reliably called in `getReturnType()`
Will push new version soon

Reply:
It was related to my previous comment:
> `Method getUInt is not supported for Tuple(UInt8)` though sounds like an internal error that we should avoid showing and probably even getting.

Reply:
Also now I remember that getReturnTypeImpl() is at least sometime called with only `.type` being filled and the `.column` actually being nullptr, so it is better to override `getReturnTypeImpl(const DataTypes & arguments)`

Reply:
That's why I rely on it being a const. But I thought that columns are resolved, otherwise it's not clear why `getReturnTypeImpl(const ColumnsWithTypeAndName & arguments)` exists as `getReturnTypeImpl(const DataTypes & arguments)` is much better

Reply:
Ok that should be fixed